Document cluster mode change (#9538)
camunda/camunda-docs: docs/self-managed/components/orchestration-cluster/zeebe/operations/modes.md

---
id: modes
title: "Cluster mode"

> description: "Switch an orchestration cluster between processing and recovery mode."

Reviewer: ```suggestion
description: "Switch an Orchestration Cluster between processing and recovery mode."
```

---

> A cluster mode change transitions every broker of an orchestration cluster between processing mode and recovery mode. The mode determines whether the cluster's partitions process data or are deactivated so the cluster can be restored from a backup.

Reviewer: ```suggestion
A cluster mode change transitions every broker in an Orchestration Cluster between processing and recovery mode. The selected mode determines whether the partitions process data or remain inactive while you restore the cluster from a backup.
```

> A mode change is a cluster configuration change, similar to [cluster scaling](cluster-scaling.md). The API acknowledges the request when it accepts the change, and each broker applies the transition asynchronously.

Reviewer 2: 💭 I think it would be nice to have a section for cluster operations as they are used for many things. Not sure if cluster scaling is the right example, but it's the only one present right now.

Author: Can refactor this later if we have an aggregate page of the Cluster ops

Reviewer 2: I started putting them in the general API reference in this [PR](https://camunda.slack.com/archives/C0AQ3DJBRPZ/p1785933974960059?thread_ts=1785930668.584519&cid=C0AQ3DJBRPZ). You can do it later.

> A mode change is a cluster configuration change, like [cluster scaling](cluster-scaling.md). The request is acknowledged as soon as the change is accepted, and the transition itself is applied asynchronously on each broker.

Reviewer: ```suggestion
A mode change is a cluster configuration change, similar to [cluster scaling](cluster-scaling.md). The API acknowledges the request when it accepts the change, and each broker applies the transition asynchronously.
```

## Cluster modes

| Mode         | Behavior                                                                                                                        |
| ------------ | ------------------------------------------------------------------------------------------------------------------------------- |
| `PROCESSING` | The default broker operational mode.                                                                                            |
| `RECOVERING` | All partitions are deactivated. No processing takes place, and only a restricted set of read-only operations remains available. |

## What happens on a broker in recovery mode

> In recovery mode, the broker stops processing and registers only a limited set of services.

Reviewer 2: 🔧 this section feels a bit a repetition of the above one, not sure if we can just add some extra info in the table 🤷

> In recovery mode there is no processing at all, and only a partial set of services is registered on the broker.

Reviewer: ```suggestion
In recovery mode, the broker stops processing and registers only a limited set of services.
```

> - **Partitions are deactivated.** Each local partition is registered as `inactive` and does not join its Raft group. No leader is elected, and no records are processed, replicated, or exported.

Reviewer: ```suggestion
- Each local partition is registered as `inactive` and does not join its Raft group. The cluster does not elect a leader or process, replicate, or export records.
```

> - **Only a partial set of services is registered.** Instead of the full partition installation, each broker starts a reduced set of services.

Reviewer: ```suggestion
- Each broker starts a reduced set of services instead of the full partition installation.
```

> ### What can be done in recovery mode

Reviewer: ```suggestion
### Operations available in recovery mode
```

> - Cluster state querying, topology requests

Reviewer: ```suggestion
- Query the cluster state and topology.
```

> - Primary storage backup store querying

Reviewer: ```suggestion
- Query the primary storage backup store.
```

> - Primary storage restore from a backup

Reviewer: ```suggestion
- Restore primary storage from a backup.
```

## Change the cluster mode

> Send a `PATCH` request to the `/mode` endpoint of the [Orchestration Cluster REST API](/apis-tools/orchestration-cluster-api-rest/specifications/change-cluster-mode.api.mdx). The endpoint uses the Orchestration Cluster REST API port, which is `8080` by default.

Reviewer 2: 🔧 we need to add the required set of permissions to be able to run this operation in this guide

Author: The OpenAPI spec from the monorepo will include the Authz in the API view automatically when added I think.

Reviewer 2: yeah it  would be nice to at least mention it?

> Send a `PATCH` request to the `/mode` endpoint of the [Orchestration Cluster REST API](/apis-tools/orchestration-cluster-api-rest/specifications/change-cluster-mode.api.mdx), available on the REST API port of the Orchestration Cluster (`8080` by default).

Reviewer: ```suggestion
Send a `PATCH` request to the `/mode` endpoint of the [Orchestration Cluster REST API](/apis-tools/orchestration-cluster-api-rest/specifications/change-cluster-mode.api.mdx). The endpoint uses the Orchestration Cluster REST API port, which is `8080` by default.
```

To enter recovery mode:

```bash
curl -X PATCH 'http://localhost:8080/v2/mode?mode=RECOVERING' \
  -H 'Accept: application/json'
```

To return to processing mode:

```bash
curl -X PATCH 'http://localhost:8080/v2/mode?mode=PROCESSING' \
  -H 'Accept: application/json'
```

### Request parameters

| Parameter | Required | Description                                                                                                           |
| --------- | -------- | --------------------------------------------------------------------------------------------------------------------- |
| `mode`    | Yes      | The target mode, either `RECOVERING` or `PROCESSING`.                                                                 |

> | `dryRun`  | No       | If `true`, the request is only validated and the resulting plan is returned without applying it. Defaults to `false`. |

Reviewer: ```suggestion
| `dryRun`  | No       | If `true`, validates the request and returns the resulting plan without applying the change. Defaults to `false`. |
```

### Response

A successful request returns `200` with the ID of the triggered cluster change and the ordered list of operations that will be applied.

> Use `dryRun=true` to review this plan before applying it:

Reviewer: ```suggestion
Use `dryRun=true` to review the change plan before applying it:
```

```bash
curl -X PATCH 'http://localhost:8080/v2/mode?mode=RECOVERING&dryRun=true' \
  -H 'Accept: application/json'
```

## Monitor the transition

> The mode change request returns before the transition has completed, so track its progress after sending it.

Reviewer: ```suggestion
The mode change request returns before the transition completes. Track the transition after you send the request.
```

Query the [cluster topology](/apis-tools/orchestration-cluster-api-rest/specifications/get-topology.api.mdx) to see the state of each broker's partitions:

```bash
curl 'http://localhost:8080/v2/topology'
```

> While a broker is in recovery mode, its partitions report `role: inactive` and `state: recovering`. After the cluster returns to processing mode, partitions report `role: leader/follower` and `state: active`.

Reviewer: ```suggestion
While a broker is in recovery mode, its partitions report `role: inactive` and `state: recovering`. After the cluster returns to processing mode, each partition reports `role: leader` or `role: follower` and `state: active`.
```

You can also query the cluster management API on the management port (`9600` by default) to follow the change by its ID:

> ```bash

Reviewer 2: 💭 it would be nice to link to a general "API Reference" here 😆

Author: If only we had one :smiley:

curl 'http://localhost:9600/orchestration/actuator/cluster'
```

> ## Considerations

Reviewer: ```suggestion
## Cluster mode change considerations
```

- Entering recovery mode stops all processing in the cluster. Plan the change as a maintenance operation, and expect client requests to fail while the cluster is recovering.

> - Brokers that are already in the target mode are not included in the plan, so repeating a request that has already been applied results in an empty plan.

Reviewer: ```suggestion
- Brokers already in the target mode are not included in the plan. Repeating the same request after the change completes results in an empty plan.
```

> - Mode change is considered a cluster configuration change, meaning that only one operation can be in-flight. You can always cancel a cluster configuration change

Reviewer: ```suggestion
- A mode change is a cluster configuration change, so only one cluster configuration operation can be in progress at a time. You can cancel an active cluster configuration change.
```
